### PR TITLE
Sharing is scaring article

### PR DESCRIPTION
**Pull request by franz1981**
This is the first draft of the article, which still need to check if the way I created code snippet/links/images/assembly is fine with adoc.

Go at https://github.com/franz1981/redhatperf.github.io/blob/sharing_is_scaring/content/post/method-data-scalability/index.adoc

### CODE REVIEW

**Comment by franz1981** [comment]
There are still few snippet and images links to fix

**Comment by franz1981** [comment]
Waiting JDK team review for the MethodData role

**Comment by franz1981** [comment]
In the second part I have to explain why https://github.com/openjdk/jdk/blob/84240cc8e01663196cf883efec4c0eb4ea590c27/src/hotspot/share/compiler/compilationPolicy.cpp#L1039-L046 doesn't save the day, will be fun 😊

And this is what copilot think about it
> The `CompilationPolicy::common` function defines the state transitions for method compilation levels in the JDK. Here are the key transitions:
> 1. **0 -> 3 -> 4**: The most common path where a method starts in the interpreter (`CompLevel_none`), gathers full profiling data (`CompLevel_full_profile`), and then is optimized by the highest level compiler (`CompLevel_full_optimization`).
> 2. **0 -> 2 -> 3 -> 4**: This occurs when the load on the high-level compiler (e.g., C2) is high. The method transitions to a limited profiling state (`CompLevel_limited_profile`) before gathering full profiling data and then being fully optimized. This helps manage the compilation load.
> 3. **0 -> (3->2) -> 4**: If the C1 queue is long, a method might gather full profiling data at level 0 and then switch to limited profiling while still in the queue, which allows faster execution while waiting for C2 compilation.
> 4. **0 -> 3 -> 1 or 0 -> 2 -> 1**: After initial compilation with C1, a method might be identified as trivial and compiled at a simpler level (`CompLevel_simple`). This can also happen if a method can't be compiled with the highest level but can be with C1.
> 5. **0 -> 4**: This path can occur if a method fails C1 compilation or due to a deoptimization that doesn't require re-profiling.
> These transitions help manage the compilation workload and optimize performance based on method characteristics and compilation queue lengths.

**Comment by franz1981** [comment]
I need to quote https://developers.redhat.com/articles/2023/09/29/how-we-solved-hotspot-performance-puzzle as well since it's very relevant

**Comment by franz1981** [comment]
I think I should add some notes on why the counters kept getting updated even if they pass the thresholds (invocation and back edge counters, really - for the others doesn't make sense to stop collecting)

**Comment by franz1981** [comment]
I need to link:
- https://mail.openjdk.org/pipermail/hotspot-dev/2024-December/099863.html
- https://developers.redhat.com/articles/2021/11/18/runtime-profiling-openjdks-hotspot-jvm#

with the former actually performing the same exact analysis I made - but with AARCH64 - and reaching the same conclusions (which is great - since it validates them!).

**Comment by franz1981** [comment]
Here the paper content, publicly accessible: https://ckirsch.github.io/publications/proceedings/MPLR24.pdf#page=117

**Comment by franz1981** [comment]
Here we go @stalep now looking how the site looks like :pray:

**Review by stalep — DISMISSED** [review]
Looks very good, almost too short :)
One thought was that the ending was a bit short on tips on how to avoid this issue. It's noted, but fairly short. Idk, might be very good as it is.

**Comment by franz1981** [comment]
Thanks @stalep Yep, I've decided to cut it short to make more clear to more people what it is - even if not super technical
Let's say that is an experiment - if people want more assembly, I'm all in eheh :dagger: 

PTAL @stalep I've updated the date

and I could add a 

### How you can fix it?

There are several ways to fix it, based on which platform the application is running on and at which level is addressed.
If users want to reduce the multi-node scalability impacts, in Openshift is possible to configure a [Topology Manager](https://docs.redhat.com/en/documentation/openshift_container_platform/4.3/html/scalability_and_performance/using-topology-manager) and save pods to run with cpu resources shared among different nodes, constraining it to run on a single one. 
Other orchestration mechanisms have similar deployment configuration options.

If users are interested into fixing the root cause which lead to run the application code in Tier 3 unoptimized compilation level (prone to the scalability issue mentioned so far) - they just need to increase the cpu quota/resources enough to sustain the expected peak load whilst performing full JIT compilation (including the more costly Tier 4 C2 one).
In short: know your workload and tune the runtime to not have long compilation pauses due to starved runtime compiler threads!


@stalep it looks complex again to me...mmmm

File: content/post/method-data-scalability/index.adoc [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+
+Tiered compilation in the HotSpot JVM balances application startup speed and runtime performance by using https://developers.redhat.com/articles/2021/06/23/how-jit-compiler-boosts-java-performance-openjdk[multiple levels] of code execution and optimization.
+Initially, it uses an **interpreter** for immediate execution. As methods are invoked frequently, it employs a fast compiler i.e. **C1** to generate native code.
+Over time, methods that are heavily used ("hot spots") are further optimized with the optimizing compiler i.e. **C2**, which applies advanced optimizations for maximum performance.
+
+This tiered approach ensures quick application responsiveness while progressively optimizing performance-critical code paths. The name "HotSpot" reflects this focus on dynamically identifying and optimizing hot spots in code execution for efficiency.
+
+What’s less known about tiered compilation is that the C2 compiler can be very CPU intensive and, when it doesn’t have enough resources, its activity https://jpbempel.github.io/2020/05/22/startup-containers-tieredcompilation.html[affects startup time].
```
Comment (lampajr) [inline content/post/method-data-scalability/index.adoc]: Just a minor but if I am not wrong the C2 compilation could also affect runtime performance and not only the startup, if so I think it would be good to clarify that here
```suggestion
What’s less known about tiered compilation is that the C2 compiler can be very CPU intensive and, when it doesn’t have enough resources, its activity could affect application performances including https://jpbempel.github.io/2020/05/22/startup-containers-tieredcompilation.html[startup time].
```

**Review by lampajr — DISMISSED** [review]
That's great @franz1981!
I found it very clear and really focused on the topic w/o getting sidetracked by other secondary issues 🚀 

I left a very minor suggestion, up to you whether it makes sense or not

**Comment by franz1981** [comment]
FYI @diegolovison This was the original commit: https://github.com/franz1981/redhatperf.github.io/blob/654fd8a50915b499f3547a50f94f97614a750153/content/post/method-data-scalability/index.adoc

**Comment by franz1981** [comment]
@stalep @lampajr if we're ready to go, can you merge it? 
Here it says "Review required" so still missing something :)

**Review by lampajr — APPROVED** [review]

**Comment by lampajr** [comment]
> @stalep @lampajr if we're ready to go, can you merge it? Here it says "Review required" so still missing something :)

Approved and merged ✅
